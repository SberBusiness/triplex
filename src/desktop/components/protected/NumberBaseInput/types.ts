import {TestProps} from '@sberbusiness/triplex/desktop/common/types/CoreTypes';
import {IInputProps} from '../../Input/Input';

/**
 * Свойства компонента NumberBaseInput.
 * @extends Omit<IInputProps, 'onChange' | 'placeholder'>, TestProps.
 */
export interface INumberBaseInputProps extends Omit<IInputProps, 'onChange' | 'max' | 'min'>, TestProps {
    /**
     * Значение.
     */
    value: string;

    /**
     * Обработчик изменения значения. Вызывается непосредственно перед onBlur (так исторически сложилось).
     */
    onChange: (value: string) => void;
    /**
     * Количество знаков в дробной части. При значении 0 подразумевается целочисленное значение.
     */
    valueFractionLength?: number;
    /**
     * Количество знаков в дробной части. При значении 0 подразумевается целочисленное значение.
     */
    placeholderFractionLength?: number;
}
