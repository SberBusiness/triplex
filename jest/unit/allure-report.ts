import {JasmineAllureReporter} from 'allure-jasmine';
import {Status, TestResult} from 'allure-js-commons';

const reporter = new JasmineAllureReporter({
    resultsDir: 'tests/out/unit/allure/allure-results',
    testMapper: (result: TestResult) => {
        if (result.status == Status.SKIPPED) result.fullName = `(WAS SKIPPED) ${result.fullName}`;
        return result;
    },
});

// @ts-ignore
jasmine.getEnv().addReporter(reporter);

export const allure = reporter.getInterface();
