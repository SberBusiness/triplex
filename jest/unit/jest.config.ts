import type {Config} from '@jest/types';

// Конфигурация Jest для запуска unit тесов.
const config: Config.InitialOptions = {
  preset: 'ts-jest',
  rootDir: '../../',
  roots: [
    '<rootDir>/src/',
    '<rootDir>/tools/',
  ],
  setupFilesAfterEnv: [
    '<rootDir>/jest/jest.setup.ts',
    '<rootDir>/jest/unit/allure-report.ts',
  ],
  moduleDirectories: [
    'node_modules',
    'src'
  ],
  moduleNameMapper: {
    "@jest/(.*)": "<rootDir>/jest/$1.ts",
    "^@sberbusiness/triplex/(.*)": "<rootDir>/src/$1"
  },
  verbose: true,
};

export default config;