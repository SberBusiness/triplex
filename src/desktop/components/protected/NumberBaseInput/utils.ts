import {formatAmount} from '@sberbusiness/triplex/desktop/utils/amountUtils';
import {AmountConst} from '@sberbusiness/triplex/desktop/common/consts/AmountConst';

/**
 * Вычисляем позицию каретки.
 *
 * @param {string} prevValueFormatted Прошлое отформатированное значение.
 * @param {string} newValueRaw Новое не отформатированное значение.
 * @param {number} fractionLength Количество знаков в дробной части. При значении 0 подразумевается целочисленное значение.
 * @param {number} currentCaretPosition Позиция каретки в поле ввода.
 */
export const countCaret = (prevValueFormatted: string, newValueRaw: string, fractionLength: number, currentCaretPosition: number) => {
    const spacesInOldValue = (prevValueFormatted.match(/\s/g) || []).length;
    const spacesInNewValue = (formatAmount(newValueRaw, fractionLength).match(/\s/g) || []).length;
    let caretSpaceOffset = spacesInNewValue - spacesInOldValue;

    if (prevValueFormatted.length - newValueRaw.length > 1) {
        // Если было удалено больше одного символа. Делаем reset caretSpaceOffset.
        caretSpaceOffset = 0;
    } else if (newValueRaw.substr(0, 1) === AmountConst.GroupSeparator) {
        // Если впереди пробел (" 786 673") сдвигаем каретку.
        caretSpaceOffset++;
    }

    return currentCaretPosition + caretSpaceOffset;
};
