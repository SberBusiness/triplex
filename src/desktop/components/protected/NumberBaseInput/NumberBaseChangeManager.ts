import * as React from 'react';
import {AmountConst} from '@sberbusiness/triplex/desktop/common/consts/AmountConst';
import {StringUtils} from '@sberbusiness/triplex/desktop/utils/stringUtils';
import {formatAmount} from '@sberbusiness/triplex/desktop/utils/amountUtils';
import {getCaretPosition} from '@sberbusiness/triplex/desktop/utils/inputUtils';

/** Возвращаемая модель данных. */
export interface IAmountBaseChangeResult {
    /** Отформатированное значение. */
    formattedValue: string;
    /** Позиция каретки. */
    caret: number;
    /** Предыдущая позиция каретки. */
    prevCaret: number;
}

/** Логика изменения значения и позиции каретки у поля Amount. */
export class NumberBaseChangeManager {
    /**
     * Обработчик изменения значения.
     * @param event Событие изменения.
     * @param prevValue Прошлое отформатированное значение.
     * @param fractionLength Количество знаков в дробной части.
     * @param maxLength Общее количество знаков.
     */
    public static handleChange(
        event: React.ChangeEvent<HTMLInputElement>,
        prevValue: string,
        fractionLength: number,
        maxLength: number
    ): IAmountBaseChangeResult {
        let value = event.target.value;
        let caret = getCaretPosition(event.target);

        const diffLength = prevValue.length - value.length;
        // Удалён один пробел.
        if (diffLength == 1 && prevValue[caret] == AmountConst.GroupSeparator) {
            return {formattedValue: prevValue, caret: caret, prevCaret: caret};
        }

        if (caret > 0) {
            const {string, caretShift} = NumberBaseChangeManager.filteringStage(value, caret, fractionLength, maxLength);
            value = string;
            caret += caretShift;
        }

        // Отсутствует значение.
        if (value == '') {
            return {formattedValue: '', caret: 0, prevCaret: diffLength};
        }

        if (fractionLength > 0) {
            value = NumberBaseChangeManager.restoringStage(value, caret, fractionLength);
        }

        const separatorIndex = value.indexOf(AmountConst.DecimalSeparator);
        // Удалены все ненулевые значения числа с дробной частью.
        if (separatorIndex == 0 && !/[1-9]/g.test(value)) {
            return {formattedValue: '', caret: 0, prevCaret: diffLength};
        }

        const {string: formattedValue, caretShift} = NumberBaseChangeManager.formattingStage(value, caret, fractionLength, separatorIndex);

        return {
            formattedValue,
            caret: caret + caretShift,
            prevCaret: caret + (formattedValue == prevValue ? caretShift : diffLength),
        };
    }

    /** Этап очищения значения от невалидных и неуместных символов. */
    private static filteringStage = (value: string, caret: number, fractionLength: number, maxLength: number) => {
        const [beforeCaret, afterCaret] = [value.substring(0, caret), value.substring(caret)];
        // Определяем индекс десятичного разделителя, необходимо для определения логики фильтрации.
        const separatorIndex = Math.max(
            value.lastIndexOf(AmountConst.DecimalSeparator),
            value.lastIndexOf(AmountConst.DecimalSeparatorCalc)
        );

        // Максимальное количество цифр, которое сможет содержать отфильтрованный фрагмент.
        let digitsLimit: number;
        // Может ли отфильтрованный фрагмент содержать десятичный разделитель.
        let separatorAllowed: boolean;

        // Если значение является целым числом.
        if (fractionLength == 0) {
            digitsLimit = maxLength - StringUtils.countDigits(afterCaret);
            separatorAllowed = false;
        }
        // Если отсутствует десятичный разделитель.
        else if (separatorIndex == -1) {
            digitsLimit = maxLength - fractionLength - AmountConst.DecimalSeparator.length;
            separatorAllowed = false;
        }
        // Если каретка перед числом целой части.
        else if (caret < separatorIndex) {
            digitsLimit = maxLength - StringUtils.countDigits(afterCaret) - AmountConst.DecimalSeparator.length;
            separatorAllowed = false;
        }
        // Если каретка перед числом дробной части.
        else if (caret > separatorIndex) {
            digitsLimit = maxLength - AmountConst.DecimalSeparator.length;
            separatorAllowed = true;
        }
        // Если каретка перед десятичным разделителем.
        else {
            digitsLimit = maxLength - StringUtils.countDigits(afterCaret) - AmountConst.DecimalSeparator.length;
            separatorAllowed = true;
        }

        const fragment = NumberBaseChangeManager.getFilteredFragment(beforeCaret, digitsLimit, separatorAllowed, !!afterCaret.length);

        return {string: fragment + afterCaret, caretShift: fragment.length - beforeCaret.length};
    };

    /** Функция, возвращающая отфильтрованную часть значения (до каретки). */
    private static getFilteredFragment = (fragment: string, digitsLimit: number, separatorAllowed: boolean, partial: boolean) => {
        const isDigit = (char: string) => char >= '0' && char <= '9';
        const isDecimalSeparator = (char: string) => char == AmountConst.DecimalSeparator || char == AmountConst.DecimalSeparatorCalc;

        let string = '';

        for (let i = 0, len = fragment.length, count = 0; i < len; i++) {
            if (isDigit(fragment[i])) {
                if (count < digitsLimit) {
                    string += fragment[i];
                    count++;
                }
            } else if (isDecimalSeparator(fragment[i])) {
                if (separatorAllowed && string.indexOf(AmountConst.DecimalSeparator) == -1) {
                    string += AmountConst.DecimalSeparator;
                }
            }
        }

        // Убираем ведущие нули.
        string = string.replace(/^0+(\d)/, '$1');

        // Проверяем, можно ли пропустить введённый ноль.
        if (string == '0' && !separatorAllowed && partial) {
            string = '';
        }

        return string;
    };

    /** Этап восстановления значения, возвращаем удаленные символы. */
    private static restoringStage = (value: string, caret: number, fractionLength: number) => {
        let string = value.substring(0, caret);

        const separatorIndex = value.indexOf(AmountConst.DecimalSeparator);
        // Если был удален десятичный разделитель, добавляем его.
        if (separatorIndex == -1) {
            string += AmountConst.DecimalSeparator;
        }

        const fractionalPart = separatorIndex !== -1 ? value.substring(separatorIndex + 1) : value.substring(caret);
        // Если в дробную часть числа добавилось значение.
        if (fractionalPart.length > fractionLength) {
            caret += fractionalPart.length - fractionLength;
        }
        // Если есть необходимость в пагинации нулями.
        else if (fractionalPart.length < fractionLength) {
            string += Array(fractionLength - fractionalPart.length + 1).join('0');
        }

        string += value.substring(caret);

        return string;
    };

    /** Этап форматирования значения к нужному виду. */
    private static formattingStage = (value: string, caret: number, fractionLength: number, separatorIndex: number) => {
        let string = '';
        let caretShift = 0;

        string = formatAmount(value, fractionLength);

        if (caret > 0) {
            caretShift = StringUtils.countWhitespaces(string) - StringUtils.countWhitespaces(value);
        }
        // Удалена целая часть числа.
        else if (separatorIndex == 0) {
            caretShift = 1;
        }

        return {string, caretShift};
    };
}
