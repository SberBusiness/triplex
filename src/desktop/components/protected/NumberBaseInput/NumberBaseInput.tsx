import React from 'react';
import {Input} from '@sberbusiness/triplex/desktop/components/Input/Input';
import {AmountConst} from '@sberbusiness/triplex/desktop/common/consts/AmountConst';
import {formatAmount} from '@sberbusiness/triplex/desktop/utils/amountUtils';
import {setCaretPosition} from '@sberbusiness/triplex/desktop/utils/inputUtils';
import {INumberBaseInputProps} from './types';
import {NumberBaseChangeManager} from './NumberBaseChangeManager';

/** Максимальная длина поля ввода суммы. 16 символов до запятой, самая запятая, 2 символа после запятой. */
const NUMBER_INPUT_MAX_LENGTH = 19;

/** Состояния NumberBaseInput. */
export interface INumberBaseInputState {
    /** Отформатированное значение. */
    formattedValue: string;
}

/** Компонент ввода суммы. */
export class NumberBaseInput extends React.Component<INumberBaseInputProps, INumberBaseInputState> {
    public static displayName = 'NumberBaseInput';

    public static defaultProps = {
        valueFractionLength: 2,
        maxLength: NUMBER_INPUT_MAX_LENGTH,
    };

    public state = {
        formattedValue: formatAmount(this.props.value, this.props.valueFractionLength),
    };

    private caret = 0;

    /**
     * Ссылка на поле ввода суммы.
     */
    private input: HTMLInputElement | undefined;

    public componentDidUpdate(prevProps: Readonly<INumberBaseInputProps>): void {
        const {value, valueFractionLength} = this.props;

        if (value !== prevProps.value || valueFractionLength !== prevProps.valueFractionLength) {
            this.setState({formattedValue: formatAmount(value, valueFractionLength)}, () => {
                this.input!.value = this.state.formattedValue;
                setCaretPosition(this.input, this.caret);
            });
        }
    }

    public render(): JSX.Element {
        const {error, placeholderFractionLength, value, valueFractionLength, title, placeholder, maxLength, ...restProps} = this.props;
        const {formattedValue} = this.state;

        return (
            <Input
                {...restProps}
                value={formattedValue}
                title={title || formattedValue}
                placeholder={placeholder || formatAmount('0', placeholderFractionLength)}
                onChange={this.handleChange}
                autoComplete="off"
                error={error}
                ref={this.setInputRef}
            />
        );
    }

    /**
     * Обработчик onChange input-а.
     * @param {React.ChangeEvent<HTMLInputElement>} event Событие.
     */
    private handleChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
        const {valueFractionLength, maxLength} = this.props;
        const {formattedValue: prevFormattedValue} = this.state;
        const field = event.target;

        const {formattedValue, caret, prevCaret} = NumberBaseChangeManager.handleChange(
            event,
            prevFormattedValue,
            valueFractionLength!,
            maxLength!
        );

        const formattedValueCalc = formatAmount(formattedValue, valueFractionLength, false, true).replace(
            AmountConst.DecimalSeparator,
            AmountConst.DecimalSeparatorCalc
        );

        field.value = prevFormattedValue;
        setCaretPosition(field, prevCaret);

        this.caret = caret;
        this.props.onChange(formattedValueCalc);
    };

    /**
     * Установка ссылки на поле ввода.
     * @param {HTMLInputElement} ref Ссылка на элемент.
     */
    private setInputRef = (ref: HTMLInputElement) => {
        this.input = ref;
    };
}
