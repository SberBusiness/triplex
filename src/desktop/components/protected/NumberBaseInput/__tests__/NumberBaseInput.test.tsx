import React from 'react';
import renderer from 'react-test-renderer';
import {NumberBaseInput} from '../NumberBaseInput';
import {allure} from '@jest/unit/allure-report';

describe('NumberBaseInput', () => {
    beforeEach(() => {
        allure.feature('AmountBaseInput');
    });

    it('renders correctly', () => {
        const component = <NumberBaseInput value="-8964152,37" onChange={jest.fn} />;
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
