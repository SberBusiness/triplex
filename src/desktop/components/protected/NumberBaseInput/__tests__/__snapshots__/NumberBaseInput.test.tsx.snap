// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberBaseInput renders correctly 1`] = `
<input
  autoComplete="off"
  className="cssClass[input]"
  onChange={[Function]}
  placeholder="0,00"
  title="-8 964 152,37"
  value="-8 964 152,37"
/>
`;
